include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,214 @@
+
+
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * pnode describes a tree of properties.
+ *
+ * A pnode can either be empty, hold a value (a string, integer, real, or bool),
+ * contain an array of pnode., or contain a mapping between strings and pnodes.
+ */
+class pnode {
+public:
+    using key_type = std::string;
+    using map_type = std::map<key_type, pnode>;
+    using array_type = std::vector<pnode>;
+
+    /**
+     * tag_t is the indicator for the current node storage.
+     */
+    enum class tag_t { empty, array, boolean, real, integer, string, map };
+
+    /**
+     * Default constructor: create an empty node
+     */
+    explicit pnode();
+
+    /**
+     * Constructor for bool
+     *
+     * @param boolean  the bool type value
+     */
+    explicit pnode(bool boolean);
+
+    /**
+     * Constructor for integer with all integer type
+     *
+     * @tparam T  input type
+     *
+     * @param integer  the integer type value
+     */
+    template <typename T,
+              std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    explicit pnode(T integer);
+
+    /**
+     * Constructor for string
+     *
+     * @param str  string type value
+     */
+    explicit pnode(const std::string& str);
+
+    /**
+     * Constructor for char* (otherwise, it will use bool)
+     *
+     * @param str  the string like "..."
+     */
+    explicit pnode(const char* str);
+
+    /**
+     * Constructor for double (and also float)
+     *
+     * @param real  the floating point type value
+     */
+    explicit pnode(double real);
+
+    /**
+     * Constructor for array
+     *
+     * @param array  an pnode array
+     */
+    explicit pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    explicit pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node tag.
+     *
+     * @return the tag
+     */
+    tag_t get_tag() const;
+
+    /**
+     * Access the array stored in this property node. Throws
+     * `gko::InvalidStateError` if the property node does not store an array.
+     *
+     * @return the array
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Access the map stored in this property node. Throws
+     * `gko::InvalidStateError` if the property node does not store a map.
+     *
+     * @return the map
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Access the boolean value stored in this property node. Throws
+     * `gko::InvalidStateError` if the property node does not store a boolean
+     * value.
+     *
+     * @return the boolean value
+     */
+    explicit bool get_boolean() const;
```

Comment:
I meant `operator bool()` above, this one is fine
```suggestion
   bool get_boolean() const;
```

Reply:
Hmm. no idea what's happened to me